src/Solvers/conjugate_gradient_poisson_solver.jl

```diff
@@ -136,15 +140,19 @@ const FFTBasedPreconditioner = RegularizedPreconditioner{<:SolverWithFFT}
 function precondition!(p, regularized::FFTBasedPreconditioner, r, args...)
     solver = regularized.preconditioner
     compute_preconditioner_rhs!(solver, r)
-    p = solve!(p, solver)
+    solve!(p, solver)
+    regularize_poisson_solution!(p, regularized)
+    return p
+end
 
+function regularize_poisson_solution!(p, regularized)
     δ = regularized.regularization
     rhs = regularized.rhs
     mean_p = mean(p)
```

Comment:
I may have found why the solver blows up for positive `regularization`.

Here, `p` is the solution in the whole grid by the FFT Poisson solver, but `mean_p` calculates the mean of `p` in the immersed boundary grid. In my test, `mean_p` is usually 1e-5, which is much larger than round-off errors. I also calculated the eigenvalues of the preconditioner with `regularization=0`, and I found both negative and positive eigenvalues.

Removing the subtraction of `mean_p` enables convergence for positive `regularization` according to my test.

Reply:
That is interesting! I'll test it. I don't completely understand why this is what's needed, but let's think about it.

Reply:
I'm still getting NaNs. What `regularization` did you use? Did you make any other changes?

Reply:
I tested the code in a simpler case without `FlatExtrapolationOpenBoundaryCondition`. That may explain why the result is different. Besides, I usually check the distribution of eigenvalues to judge whether the solver is numerically stable than running the simulation. I am looking into the case with `FlatExtrapolationOpenBoundaryCondition` and aim to understand how it changes the distribution of eigenvalues.

Reply:
Just to write out the idea, we are trying to solve a "regularized" Poisson equation:

$$ \tilde L \phi = R $$

where $\tilde L = \nabla^2 - \delta V^{-1} \int \mathrm{d V}$$ is the regularized Laplace operator (on an immersed boundary grid) with the singularity removed.

We're proposing to use a preconditioner that inverts the "non-immersed" (but regularized) Laplace operator,

$$ L_u = \nabla_u^2 - \delta V_\star^{-1} \int_{\Omega_\star} \mathrm{d V} $$

where $\nabla_u^2$ applies on the "underlying grid" (ie ignoring the immersed boundary). The present discussion is how to define the mean, which is taken over the domain $\Omega_\star$ with volume $V_\star$. The current code takes $V_\star = V$, but we can also consider taking $V_\star = V_u$ and $\Omega_\star = \Omega_u$.

Reply:
By the way @Yixiao-Zhang I think this example really meaningfully exposes the effect of regularization, because it's only with this kind of open boundary condition that $\bar R \ne 0$. I believe that with impenetrable (and periodic of course) boundary conditions $\bar R$ is very small.

Reply:
I did not realized $\overline{R} \neq 0$ for this kind of open boundary condition. However, I think the current pressure solver cannot handle `FlatExtrapolationOpenBoundaryCondition` properly. As far as I understand, the pressure solver is unaware of the open boundary condition, but the open boundary condition should affect how `fill_halo_region!` works for `rhs`. The original boundary condition for `rhs` is $\hat{n}\cdot \nabla r = 0$, but it should be  $(\hat{n}\cdot \nabla)^2 r = 0$ for open boundary conditions? \(I am not sure.\)

Right now, `rhs`  is defined in this way:

```Julia
rhs = CenterField(grid)
```

The good news is that the Laplace operator becomes negative definite if we take the open boundary into account. So, there is no need to regularize the Poisson solver. \(The preconditioner is still semi-definite, though.\)

Reply:
> the pressure solver is unaware of the open boundary condition

This isn't quite true. The inhomogeneous part of any boundary condition is incorporated into the Poisson equation RHS (this happens implicitly by 1) filling halo regions on the predictor velocity prior to the pressure solve and then 2) taking the divergence of the predictor velocity.) This is why the operator / FFT-solve must use homogeneous Neumann conditions for the algorithm to work.